best_practices.md

```diff
@@ -0,0 +1,147 @@
+# CLUES BEST PRACTICES
+
+A guide for getting the most out of Clues.
+
+## CTX
+
+### Tracing
+
+Calling `AddTrace` or `AddTraceName` will append to an internal property
+with the key `clues_trace`.  Clues traces form a slice of comma delimited
+hashes (or names, when using TraceName).  These can be useful for filtering
+logs to certain process branches.
+
+clues.Add() always automatically appends to the trace, so you don't need
+to go out of your way to get the benefits unless you specifically want to
+add or name traces.
+
+```go
+func main() {
+    ctx := clues.Add(context.Background(), "k", "v")
+    // clues_trace: abc123
+    foo(ctx)
+}
+
+func foo(ctx context.Context) {
+    ctx = clues.AddTrace(ctx)
+    // clues_trace: abc123,xyz987
+    bar(ctx)
+}
+
+func bar(ctx context.Context) {
+    ctx = clues.AddTraceName(ctx, "bar", "k2", "v2")
+    // clues_trace: abc123,xyz987,bar
+    // ...
+}
+```
+
+### Iterators
+
+Create a separate ctx variable when recording clues within a loop.
+
+```go
+ctx = clues.Add(ctx, "k", "v")
+
+for _, user := range users {
+    ictx := clues.Add(ctx, "user_id", user.ID)
+    handleUser(ictx, user)
+}
+```
+
+## ERRORS
+
+### Always Stack
+
+A single-error Stack ensures clues will append a stacktrace reference
+for that point of return.  New(), Wrap() and Stack()ing multiple errors
+all do this same process.  This tip is specific for cases where you'd
+normally `return err`.
+
+```go
+err := downstreamCall(ctx)
+if err != nil {
+    return clues.Stack(err)
+}
+```
+
+### With Ctx As Needed
+
+Clues doesn't automatically push the clues from a ctx into an error.  When
+you want to do that, you can either append to a clues error using WithClues(ctx),
+or call a constructor that includes the ctx.
+
+While extracting errors from a ctx multiple times is a benign process, it
```

Comment:
I think this should be `clues` instead of errors? Not completely sure so taking a guess based off the context clues 😛

```suggestion
While extracting clues from a ctx multiple times is a benign process, it
```

Reply:
Ah, totally.  the wording there is odd overall; I've given it a little facelift for clarity.

Reply:
hmmm I'm still a little confused though, we don't extract errors from the context do we? I thought we only extracted clues from the context

Reply:
hah, yep, you jumped ahead a little.  I meant, generally, "put ctx values into the error".